src/classes/RD2_ScheduleService_TEST.cls

```diff
@@ -241,6 +241,60 @@ private with sharing class RD2_ScheduleService_TEST {
 
     }
 
+    /***
```

Comment:
What might be helpful in this test class is to create unit tests that explicitly validate the results of the `getNextDonationDate` method without going through the full service. That will result is smaller test methods that can be easier to manipulate and create multiple variations for. For example:

```
@IsTest
private static void shouldCalculateNddWhenSomeCondition() {
    RecurringDonationSchedule__c rds =  new RecurringDonationSchedule__c(
       InstallmentPeriod__c = RD2_Constants.PERIOD_MONTHLY,
       InstallmentFrequency__c = 2,
       StartDate = Date.newInstance(2019,6,1)
    );
    RD2_ScheduleService scheduleService = new RD2_ScheduleService();
    scheduleService.currentDate = Date.newInstance(2019,11,15);
    Date nextDonationDate = scheduleService.getNextDonationDate(new List<RecurringDonationSchedule__c> { rds  });
    System.assertEquals( ... );
}
```

Reply:
There are already a bunch of those.  Are there some you see should be added?

```
    @isTest
    private static void shouldCreateNextDonationDateFromWeeklyStartDate() {

        npe03__Recurring_Donation__c rd =
            TEST_RecurringDonationBuilder.constructEnhancedBuilder()
                .withMockId()
                .withContact(UTIL_UnitTestData_TEST.mockId(Contact.SObjectType))
                .withAmount(400)
                .withPaymentMethod('Check')
                .withInstallmentPeriodWeekly()
                .withInstallmentFrequency(1)
                .withDateEstablished(Date.newInstance(2019, 9, 15))
                .withStartDate(Date.newInstance(2018, 3, 17))
                .build();

        RD2_ScheduleService service = getScheduleServiceWithCurrentDateOverride(Date.newInstance(2018, 3, 17));

        System.assertEquals(Date.newInstance(2018, 3, 17),
            service.getNextDonationDate(rd),
            'Next Donation Date should equal 3/17/2018');

    }
```

Reply:
Is your suggestion to avoid creating from RD then?

Reply:
I see your point, however there are two notes with this:

- This is not actually testing building the NDD off the Schedule Records which is important. It's always recreating the Schedules from the RD record itself - which effectively ignores any existing Schedule records and any special scheduling logic they may implement (such as skips, etc.).
- You can use `withCalculateNextDonationDate()` in the factory to have it automatically calculate the NDD and apply to the NextPaymentDate field.

Reply:
I see your point.  Sound good!  

I'll make adjustments to what is already there if that sounds okay to you, since as you said we really need to test the behavior with a collection of schedules as well.